text/0000-cfg-path-version.md

```diff
@@ -0,0 +1,484 @@
+
+# Motivation
+[motivation]: #motivation
+
+[stability_stagnation]: https://blog.rust-lang.org/2014/10/30/Stability.html
+[what_is_rust2018]: https://blog.rust-lang.org/2018/07/27/what-is-rust-2018.html
+
+A core tenant of Rust's story is
+[*"stability without stagnation"*][stability_stagnation].
+We have made great strides sticking to this story while continuously
+improving the language and the community. This is especially the case with
+the coming [Rust 2018 edition][what_is_rust2018].
+
+However, while the situation for evolving the language is doing well,
+the situation for library authors is not as good as it could be.
+Today, crate authors often face a dilemma: - *"Shall I provide more features
+and implementations for later versions of Rust, or should I stay compatible
+with more versions of the compiler"*.
+
+[cargo_version_selection]: http://aturon.github.io/2018/07/25/cargo-version-selection/
+
+While [much thought][cargo_version_selection] has been given to how we can
+reduce "dependency hell" by enhancing cargo for:
+
++ the **control** users have over their dependencies.
++ the **compatibility** of crates with each other.
++ reducing the **maintainability** burden of having to make sure that
+  versions work with each other.
+
+[RFC 2483]: https://github.com/rust-lang/rfcs/pull/2483
+
+...not much focus has been given to how you can improve the situation can be
+improved by enhancing conditional compilation to extend how many versions back
+a crate supports. This becomes critically important if and when we gain LTS
+channels as proposed by [RFC 2483].
+
+[version_check]: https://crates.io/crates/version_check
+
+The current support for such conditional compilation is lacking.
+While [it is possible][version_check] to check if you are on a nightly
+compiler or to check if you are above a certain compiler version,
+such facilities are not particularly ergonomic at the moment.
+In particular, they require the setting up of a `build.rs` file and
+declaring up-front which versions you are interested in knowing about.
+These tools are also unable to check if a certain path exists and instead
+force you to know which version they were introduced in.
+
+*We can do better.* In this RFC we aim to rectify this by giving library
+authors the tools they need in the language itself. With the features
+proposed in the [summary] we aim to make retaining *compatibility* and
+supporting more compiler versions *pain-free* and to give authors a lot
+of *control* over what is supported and what is not.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+## `#[cfg(nightly)]` and `#[cfg(path_exists($path)]`
+
+Consider for a moment that we would like to use the `Iterator::flatten`
+method of the standard library if it exists, but otherwise fall back to
+`Itertools::flatten`. We can do that with the following snippet:
+
+```rust
+#![cfg_attr(nightly, feature(iterator_flatten))]
```

Comment:
The benefit of using `nightly` is that it can be combined with other things with `not`, `any`, `all`.
However, we might be able to have some sort of `attribute_alias` feature in the future that would enable this in user code.